File: adapters/macos/Sources/kakao-macos-bridge/SelfTest.swift
import Foundation

/// `kakao-macos-bridge --self-test` — parser regression checks against the
/// bundled `scenario-basic` fixture (a hand-sanitised copy of the real
/// KakaoTalk 26.6.1 tree structure). Toolchain-independent stand-in for the
/// swift-testing suite in `Tests/`.
///
/// Exit 0 = all passed, 1 = a check failed. Run after adding a new per-version
/// fixture or touching `Parsers` / `Selectors`.
enum SelfTest {
    private static var failures = 0

    private static func check(_ condition: Bool, _ label: String) {
        print(condition ? "  ok   \(label)" : "  FAIL \(label)")
        if !condition { failures += 1 }
    }

    static func run() -> Never {
        let selectors = SelectorMap.v26_6
        // Fixed clock so "오전 11:17" / "오전 10:00" produce stable ISO strings.
        let now = ISO8601DateFormatter().date(from: "2026-08-29T00:00:00Z") ?? Date()

        guard let url = Bundle.module.url(
                forResource: "scenario-basic", withExtension: "json", subdirectory: "Fixtures"),
              let data = try? Data(contentsOf: url),
              let tree = try? JSONDecoder().decode(FixtureNode.self, from: data)
        else {
            FileHandle.standardError.write(Data("could not load scenario-basic fixture\n".utf8))
            exit(1)
        }

        let mainWindow = tree.children.first { $0.title == "카카오톡" }!
        let convWindow = tree.children.first { $0.title == "개발팀" }!

        print("scenario-basic — listRooms")
        let rooms = Parsers.rooms(in: mainWindow, selectors: selectors, now: now)
        check(rooms.count == 3, "3 rooms parsed (spacer row skipped)")
        check(rooms.map(\.title) == ["개발팀", "엄마", "개발 공지"], "titles in list order")
        check(rooms.map(\.roomId) == ["row:0", "row:1", "row:2"], "opaque roomIds are row indices")
        check(rooms[0].unreadCount == 2, "bare numeric AXStaticText -> unread 2")
        check(rooms[1].unreadCount == 0, "no unread badge -> 0")
        check(rooms[0].memberCount == 18, "Count Label -> member count 18")
        check(rooms[1].memberCount == 2, "no Count Label -> 1:1 -> member count 2")
        check(rooms[2].memberCount == 42, "group Count Label -> 42")
        check(rooms[0].lastMessage?.text == "배포 끝났어요?", "last message preview from AXTextArea")
        check(rooms[2].lastMessage == nil, "empty preview -> nil lastMessage")
        check((rooms[0].lastMessage?.at ?? "").hasSuffix("Z"), "preview timestamp parsed to ISO UTC")

        print("scenario-basic — readRecent")
        let messages = Parsers.messages(in: convWindow, selectors: selectors, myName: "수빈", now: now)
        check(messages.count == 4, "4 messages parsed (spacer skipped)")
        check(messages[0].text == "배포 끝났어요?", "first message body")
        check(messages[0].sender == "민수", "sender from AXStaticText next to profile button")
        check(messages[0].kind == .text, "text kind")
        check(!(messages[0].at).isEmpty && messages[0].at.hasSuffix("Z"), "message time parsed to ISO")
        check(messages[1].text == "확인은요?", "second message body")
        check(messages[1].sender == "민수", "sender carried forward to a continuation row")
        check(messages[1].at == messages[0].at, "time inherited when row has no timestamp label")
        check(messages[2].sender == "수빈", "new sender run picked up")
        check(messages[2].at != messages[1].at, "multiline '1\\n오전 11:30' timestamp parsed (not inherited)")
        check(messages[2].outgoing, "sender == myName -> outgoing")
        check(messages[3].kind == .unsupported && messages[3].text == "", "media cell -> unsupported")
        check(messages[3].sender == "수빈", "sender carries to media cell")

        print("Korean time parsing")
        check(KoreanTime.parseHourMinute("오전 11:17").map { $0 == (11, 17) } ?? false, "오전 11:17 -> 11:17")
        check(KoreanTime.parseHourMinute("오후 12:12").map { $0 == (12, 12) } ?? false, "오후 12:12 -> 12:12")
        check(KoreanTime.parseHourMinute("오전 12:30").map { $0 == (0, 30) } ?? false, "오전 12:30 -> 00:30")
        check(KoreanTime.toISO("어제", now: now).hasSuffix("Z"), "어제 -> ISO date")
        check(KoreanTime.toISO("", now: now) == "", "empty label -> empty ISO")

        print("contract shape")
        let sr = SendResult(status: .unknown, at: nil, error: .sendVerifyTimeout)
        let json = (try? JSONEncoder().encode(sr)).flatMap { String(data: $0, encoding: .utf8) } ?? ""
        check(json.contains("\"status\":\"unknown\""), "SendResult status wire value")
        check(json.contains("\"error\":\"SEND_VERIFY_TIMEOUT\""), "ErrorCode wire value")
        check(ErrorCode.accessibilityPermissionDenied.rawValue == "ACCESSIBILITY_PERMISSION_DENIED",
              "ErrorCode raw values match contract")

        print("")
        if failures == 0 { print("all checks passed"); exit(0) }
        print("\(failures) check(s) failed"); exit(1)
    }
}

File: adapters/windows/src/self_test.rs
//! `kakao-windows-bridge --self-test` — parser regression checks against the
//! bundled `scenario-basic` fixture. Host-independent (runs on macOS via
//! `cargo test` too — see `tests/`). Mirrors the macOS `--self-test`.

use time::macros::datetime;

use crate::node::FixtureNode;
use crate::{parsers, selectors};

const FIXTURE: &str = include_str!("../fixtures/scenario-basic.json");

pub fn run() -> ! {
    let failures = run_checks(|label, ok| {
        println!("  {} {label}", if ok { "ok  " } else { "FAIL" });
    });
    if failures == 0 {
        println!("\nall checks passed");
        std::process::exit(0);
    }
    println!("\n{failures} check(s) failed");
    std::process::exit(1);
}

/// Returns the number of failed checks. `report` is called per check.
pub fn run_checks(mut report: impl FnMut(&str, bool)) -> usize {
    let mut failures = 0;
    let mut check = |label: &str, cond: bool| {
        report(label, cond);
        if !cond {
            failures += 1;
        }
    };

    let now = datetime!(2026-08-29 00:00:00 UTC);
    let sel = selectors::for_version(None).unwrap();
    let tree: FixtureNode = serde_json::from_str(FIXTURE).expect("fixture parses");

    let main = tree
        .children
        .iter()
        .find(|w| w.name.as_deref() == Some("카카오톡"))
        .expect("main window");
    let conv = tree
        .children
        .iter()
        .find(|w| w.name.as_deref() == Some("개발팀"))
        .expect("conversation window");

    println!("scenario-basic — listRooms");
    let rooms = parsers::rooms(main, sel, now);
    check("3 rooms parsed (spacer skipped)", rooms.len() == 3);
    check(
        "titles in list order",
        rooms.iter().map(|r| r.title.as_str()).collect::<Vec<_>>()
            == ["개발팀", "엄마", "개발 공지"],
    );
    check(
        "opaque roomIds are row indices",
        rooms.iter().map(|r| r.room_id.as_str()).collect::<Vec<_>>() == ["row:0", "row:1", "row:2"],
    );
    check("unread badge -> 2", rooms[0].unread_count == 2);
    check("no unread badge -> 0", rooms[1].unread_count == 0);
    check("member count 18 (group)", rooms[0].member_count == Some(18));
    check(
        "no member count -> 1:1 -> 2",
        rooms[1].member_count == Some(2),
    );
    check(
        "last message preview",
        rooms[0].last_message.as_ref().map(|m| m.text.as_str()) == Some("배포 끝났어요?"),
    );
    check(
        "empty preview -> no lastMessage",
        rooms[2].last_message.is_none(),
    );
    check(
        "preview timestamp -> ISO UTC",
        rooms[0]
            .last_message
            .as_ref()
            .is_some_and(|m| m.at.ends_with('Z')),
    );

    println!("scenario-basic — readRecent");
    let msgs = parsers::messages(conv, sel, Some(1000.0), now);
    check("5 messages parsed (spacer skipped)", msgs.len() == 5);
    check("first message body", msgs[0].text == "배포 끝났어요?");
    check("sender from Text next to profile", msgs[0].sender == "민수");
    check("left-aligned bubble -> incoming", !msgs[0].outgoing);
    check("message time -> ISO", msgs[0].at.ends_with('Z'));
    check("sender carried to continuation", msgs[1].sender == "민수");
    check(
        "time inherited on no-timestamp row",
        msgs[1].at == msgs[0].at,
    );
    check("outgoing message body", msgs[2].text == "넵 확인했습니다");
    check("right-aligned bubble -> outgoing", msgs[2].outgoing);
    check("outgoing carries no sender", msgs[2].sender.is_empty());
    check("new incoming sender run", msgs[3].sender == "수빈");
    check(
        "multiline timestamp parsed (not inherited)",
        msgs[3].at != msgs[1].at,
    );
    check(
        "media item -> unsupported",
        matches!(msgs[4].kind, kakao_contract::MessageKind::Unsupported) && msgs[4].text.is_empty(),
    );

    println!("korean time parsing");
    check(
        "오전 11:17 -> (11,17)",
        crate::korean_time::parse_hour_minute("오전 11:17") == Some((11, 17)),
    );
    check(
        "오후 12:12 -> (12,12)",
        crate::korean_time::parse_hour_minute("오후 12:12") == Some((12, 12)),
    );
    check(
        "오전 12:30 -> (0,30)",
        crate::korean_time::parse_hour_minute("오전 12:30") == Some((0, 30)),
    );
    check(
        "어제 -> ISO",
        crate::korean_time::to_iso("어제", now).ends_with('Z'),
    );
    check(
        "empty -> empty",
        crate::korean_time::to_iso("", now).is_empty(),
    );

    println!("contract shape");
    let sr = kakao_contract::SendResult {
        status: kakao_contract::SendStatus::Unknown,
        at: None,
        error: Some(kakao_contract::ErrorCode::SendVerifyTimeout),
    };
    let json = serde_json::to_string(&sr).unwrap();
    check(
        "SendResult status wire value",
        json.contains("\"status\":\"unknown\""),
    );
    check(
        "ErrorCode wire value",
        json.contains("\"error\":\"SEND_VERIFY_TIMEOUT\""),
    );
    check(
        "ErrorCode raw values match contract",
        kakao_contract::ErrorCode::AccessibilityPermissionDenied.as_str()
            == "ACCESSIBILITY_PERMISSION_DENIED",
    );

    failures
}
